## EDA of the ABO Dataset

In [1]:
from mediapipe.tasks import python
from mediapipe.tasks.python import text

import tarfile
import gzip
import numpy as np
import pandas as pd
from tqdm import tqdm

In [ ]:
import warnings
warnings.filterwarnings('ignore')

Read the JSON data right out of the tar file.

In [8]:
def read_json_from_tar(tar_file):
    # Open the tar file
    dflist = []
    with tarfile.open(tar_file, 'r') as tar:
        for member in tqdm(tar.getmembers()):
            if member.name.endswith('json.gz'):
                #print(f"Reading {member.name}...")
                # Extract the gz file in memory
                f = tar.extractfile(member)
                if f is not None:
                  with gzip.open(f, 'rt') as f:
                    df = pd.read_json(f, lines=True)
                    dflist.append(df)
    pdf = pd.concat(dflist).set_index('item_id')
    return pdf

pdf = read_json_from_tar("../../ABO_dataset/abo-listings.tar")

100%|██████████| 20/20 [00:29<00:00,  1.47s/it]


In [5]:
pdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 147702 entries, B07NQ437BB to B07DFWQV5V
Data columns (total 27 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   brand                147643 non-null  object
 1   bullet_point         131570 non-null  object
 2   color                116180 non-null  object
 3   fabric_type          8193 non-null    object
 4   item_name            147702 non-null  object
 5   item_weight          106194 non-null  object
 6   model_name           81579 non-null   object
 7   model_number         124091 non-null  object
 8   product_type         147702 non-null  object
 9   style                43188 non-null   object
 10  main_image_id        147127 non-null  object
 11  other_image_id       137976 non-null  object
 12  color_code           18243 non-null   object
 13  country              147702 non-null  object
 14  marketplace          147702 non-null  object
 15  domain_name          14770

### Data Filtering

Drop some of the columns we won't be using.

In [8]:
pdf = pdf.drop(columns=['model_number',
                        'color_code',
                        'node',
                        'item_dimensions',
                        'spin_id',
                        '3dmodel_id',
                        'item_shape'])
# pdf = pdf[pdf['country'] == 'US']
# pdf = pdf[pdf['domain_name'] == 'amazon.com']
pdf

brand  \
item_id                                                         
B07NQ437BB  [{'language_tag': 'fr_FR', 'value': 'Amazon Es...   
B0857LSVB7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07C5FF8QS  [{'language_tag': 'en_SG', 'value': 'AmazonBas...   
B07K591232  [{'language_tag': 'it_IT', 'value': 'AmazonBas...   
B07TG425LX  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
...                                                       ...   
B07P56ZF4T      [{'language_tag': 'en_US', 'value': 'find.'}]   
B0854LJBYD  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B081HNTNK7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07THSGHVP  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07DFWQV5V  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   

                                                 bullet_point  \
item_id                                                         
B07NQ437BB  [{'language_tag': 'fr_FR', 'value': 'Plat clas...   
B0857LSVB7  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B07C5FF8QS  [{'language_tag': 'en_SG', 'value': 'For best ...   
B07K591232                                                NaN   
B07TG425LX  [{'language_tag': 'en_IN', 'value': '3D Printe...   
...                                                       ...   
B07P56ZF4T  [{'language_tag': 'en_US', 'value': 'An Amazon...   
B0854LJBYD  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B081HNTNK7  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B07THSGHVP  [{'language_tag': 'en_IN', 'value': '3D Printe...   
B07DFWQV5V  [{'language_tag': 'hi_IN', 'value': 'नॉन-स्टिक...   

                                                        color  \
item_id                                                         
B07NQ437BB  [{'language_tag': 'fr_FR', 'standardized_value...   
B0857LSVB7  [{'language_tag': 'en_IN', 'standardized_value...   
B07C5FF8QS                                                NaN   
B07K591232                                                NaN   
B07TG425LX  [{'language_tag': 'en_IN', 'standardized_value...   
...                                                       ...   
B07P56ZF4T  [{'language_tag': 'en_US', 'standardized_value...   
B0854LJBYD  [{'language_tag': 'en_IN', 'standardized_value...   
B081HNTNK7  [{'language_tag': 'en_IN', 'standardized_value...   
B07THSGHVP  [{'language_tag': 'en_IN', 'standardized_value...   
B07DFWQV5V  [{'language_tag': 'hi_IN', 'standardized_value...   

                                                  fabric_type  \
item_id                                                         
B07NQ437BB  [{'language_tag': 'fr_FR', 'value': '100% Synt...   
B0857LSVB7                                                NaN   
B07C5FF8QS                                                NaN   
B07K591232                                                NaN   
B07TG425LX                                                NaN   
...                                                       ...   
B07P56ZF4T                                                NaN   
B0854LJBYD                                                NaN   
B081HNTNK7                                                NaN   
B07THSGHVP                                                NaN   
B07DFWQV5V                                                NaN   

                                                    item_name  \
item_id                                                         
B07NQ437BB  [{'language_tag': 'fr_FR', 'value': 'Amazon Es...   
B0857LSVB7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07C5FF8QS  [{'language_tag': 'en_SG', 'value': 'AmazonBas...   
B07K591232  [{'language_tag': 'it_IT', 'value': 'AmazonBas...   
B07TG425LX  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
...                                                       ...   
B07P56ZF4T  [{'language_tag': 'en_US', 'value': 'Amazon Br...   
B0854LJBYD  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B081HNTNK7  [{'languag

We only want to keep the rows that have an English tag.

In [10]:
def has_english_tag(row):
    for value in row:
        if isinstance(value, list) and all(isinstance(item, dict) for item in value):
            for item in value:
                if 'language_tag' in item and item['language_tag'].startswith('en_'):
                    return True
    return False

pdf = pdf[[has_english_tag(pdf.loc[item_id]) for item_id in tqdm(pdf.index)]]

100%|██████████| 147702/147702 [34:29<00:00, 71.35it/s] 


In [11]:
pdf

brand  \
item_id                                                         
B0857LSVB7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07C5FF8QS  [{'language_tag': 'en_SG', 'value': 'AmazonBas...   
B07TG425LX  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07LCHFZCW  [{'language_tag': 'en_US', 'value': 'Amazon Go'}]   
B077W2YX72  [{'language_tag': 'en_US', 'value': 'Wickedly ...   
...                                                       ...   
B08569DDNF  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B0854LJBYD  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B081HNTNK7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07THSGHVP  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07DFWQV5V  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   

                                                 bullet_point  \
item_id                                                         
B0857LSVB7  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B07C5FF8QS  [{'language_tag': 'en_SG', 'value': 'For best ...   
B07TG425LX  [{'language_tag': 'en_IN', 'value': '3D Printe...   
B07LCHFZCW                                                NaN   
B077W2YX72  [{'language_tag': 'en_US', 'value': 'One 44-ou...   
...                                                       ...   
B08569DDNF  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B0854LJBYD  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B081HNTNK7  [{'language_tag': 'en_IN', 'value': 'Snug fit ...   
B07THSGHVP  [{'language_tag': 'en_IN', 'value': '3D Printe...   
B07DFWQV5V  [{'language_tag': 'hi_IN', 'value': 'नॉन-स्टिक...   

                                                        color fabric_type  \
item_id                                                                     
B0857LSVB7  [{'language_tag': 'en_IN', 'standardized_value...         NaN   
B07C5FF8QS                                                NaN         NaN   
B07TG425LX  [{'language_tag': 'en_IN', 'standardized_value...         NaN   
B07LCHFZCW                                                NaN         NaN   
B077W2YX72                                                NaN         NaN   
...                                                       ...         ...   
B08569DDNF  [{'language_tag': 'en_IN', 'standardized_value...         NaN   
B0854LJBYD  [{'language_tag': 'en_IN', 'standardized_value...         NaN   
B081HNTNK7  [{'language_tag': 'en_IN', 'standardized_value...         NaN   
B07THSGHVP  [{'language_tag': 'en_IN', 'standardized_value...         NaN   
B07DFWQV5V  [{'language_tag': 'hi_IN', 'standardized_value...         NaN   

                                                    item_name  \
item_id                                                         
B0857LSVB7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07C5FF8QS  [{'language_tag': 'en_SG', 'value': 'AmazonBas...   
B07TG425LX  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07LCHFZCW  [{'language_tag': 'en_US', 'value': 'Amazon Ki...   
B077W2YX72  [{'language_tag': 'en_US', 'value': 'Wickedly ...   
...                                                       ...   
B08569DDNF  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B0854LJBYD  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B081HNTNK7  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07THSGHVP  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   
B07DFWQV5V  [{'language_tag': 'en_IN', 'value': 'Amazon Br...   

                                                  item_weight  \
item_id                                                         
B0857LSVB7  [{'normalized_value': {'unit': 'pounds', 'valu...   
B07C5FF8QS                                                NaN   
B07TG425LX  [{'normalized_value': {'unit': 'pounds', 'valu...   
B07LCHFZCW                                                NaN   
B077W2YX72  [{'normalized_value': {'unit': 'pounds', 'valu...   
...                                                       ...

For the remaining rows, whenever there are language tags, we only keep the English ones.

In [12]:
# Previously only took the first English item from a list. Problem if there were multiple, as in bullet points. Fixed.

def extract_value(value):
    if isinstance(value, list) and all(isinstance(item, dict) for item in value):
        english_items = []
        for item in value:
            if 'language_tag' in item and item['language_tag'].startswith('en_') \
              or 'language_tag' not in item and 'value' in item:
                english_items.append(item['value'])
        if len(english_items) == 1:
            return english_items[0]
        return english_items
    return value

for col in tqdm(pdf.columns):
    for i in range(len(pdf[col])):
        pdf[col][i] = extract_value(pdf[col][i])

100%|██████████| 20/20 [08:36<00:00, 25.81s/it]


That processing took a while, so save the result.

In [2]:
# pdf.to_pickle("../../ABO_dataset/abo-listings-english-tags.pkl")
pdf = pd.read_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-english-tags.pkl")
pdf

brand  \
item_id                             
B0857LSVB7  Amazon Brand - Solimo   
B07C5FF8QS           AmazonBasics   
B07TG425LX  Amazon Brand - Solimo   
B07LCHFZCW              Amazon Go   
B077W2YX72         Wickedly Prime   
...                           ...   
B08569DDNF  Amazon Brand - Solimo   
B0854LJBYD  Amazon Brand - Solimo   
B081HNTNK7  Amazon Brand - Solimo   
B07THSGHVP  Amazon Brand - Solimo   
B07DFWQV5V  Amazon Brand - Solimo   

                                                 bullet_point          color  \
item_id                                                                        
B0857LSVB7  [Snug fit for Samsung Galaxy J2 Ace, with perf...     Multicolor   
B07C5FF8QS  For best performance, follow the manufacturer'...            NaN   
B07TG425LX  [3D Printed Hard Back Case Mobile Cover for Hu...         Others   
B07LCHFZCW                                                NaN            NaN   
B077W2YX72  [One 44-ounce plastic tub, Contains wheat, pea...            NaN   
...                                                       ...            ...   
B08569DDNF  [Snug fit for Samsung Galaxy A2 Core, with per...     Multicolor   
B0854LJBYD  [Snug fit for iKall K200, with perfect cut-out...     Multicolor   
B081HNTNK7  [Snug fit for Mobile, with perfect cut-outs fo...  multi-colored   
B07THSGHVP  [3D Printed Hard Back Case Mobile Cover for Xi...         Others   
B07DFWQV5V  [Non-stick heating plates are double Teflon co...          Black   

           fabric_type                                          item_name  \
item_id                                                                     
B0857LSVB7         NaN  Amazon Brand - Solimo Designer Lion UV Printed...   
B07C5FF8QS         NaN  AmazonBasics High Mileage Motor Oil, Synthetic...   
B07TG425LX         NaN  Amazon Brand - Solimo Designer Semi Circle Tex...   
B07LCHFZCW         NaN   Amazon Kitchen, Cajun Style Potato Salad, 7.2 oz   
B077W2YX72         NaN  Wickedly Prime Peanut Butter-Filled Pretzels, ...   
...                ...                                                ...   
B08569DDNF         NaN  Amazon Brand - Solimo Designer Chinnese Yin an...   
B0854LJBYD         NaN  Amazon Brand - Solimo Designer Teddy Bear UV P...   
B081HNTNK7         NaN  Amazon Brand - Solimo Designer Alphabet Patter...   
B07THSGHVP         NaN  Amazon Brand - Solimo Designer Simple White Ra...   
B07DFWQV5V         NaN  Amazon Brand - Solimo Tandoor Grill (1500W, Bl...   

           item_weight              model_name                  product_type  \
item_id                                                                        
B0857LSVB7          50   Samsung Galaxy J2 Ace           CELLULAR_PHONE_CASE   
B07C5FF8QS         NaN                     NaN                      AUTO_OIL   
B07TG425LX          50          Huawei P9 lite           CELLULAR_PHONE_CASE   
B07LCHFZCW         NaN                     NaN                       GROCERY   
B077W2YX72         2.8                     NaN                       GROCERY   
...                ...                     ...                           ...   
B08569DDNF          50  Samsung Galaxy A2 Core           CELLULAR_PHONE_CASE   
B0854LJBYD          50              iKall K200           CELLULAR_PHONE_CASE   
B081HNTNK7          50            Vivo V17 Pro           CELLULAR_PHONE_CASE   
B07THSGHVP          50    Xiaomi Redmi Y1 Lite           CELLULAR_PHONE_CASE   
B07DFWQV5V         NaN              SOL_KA_103  COUNTERTOP_GRIDDLE_APPLIANCE   

                                     style main_image_id  \
item_id                                                    
B0857LSVB7                             NaN   81-DuD5XzmL   
B07C5FF8QS  High Mileage - Synthetic Blend   81YCp3dcurL   
B07TG425LX                             NaN   71ZoXfUr-sL   
B07LCHFZCW                             NaN   61bCuBtuZ4L   
B077W2YX72                             NaN   71QbaO8qZIL   
...                                    ...

#### Save a sample for testing in other notebooks

We want a somewhat representative sample, so we group by product type, then save 10 or the number of products in the group, whichever is less.

In [14]:
pdf.groupby('product_type').size()

product_type
ABIS_BEAUTY              16
ABIS_BOOK                 3
ABIS_DRUGSTORE           73
ABIS_ELECTRONICS         11
ABIS_HOME_IMPROVEMENT     8
                         ..
WOUND_DRESSING            9
WRENCH                   38
WRITING_BOARD            26
WRITING_INSTRUMENT       78
WRITING_PAPER             1
Length: 529, dtype: int64

In [17]:
def sample_groups(group, rng):
    n = min(len(group), 10)
    return group.sample(n=n, random_state=rng)
    
rng = np.random.default_rng(seed=42)
pdf_sample = pdf.groupby('product_type').apply(lambda group: sample_groups(group, rng)).reset_index(level='product_type', drop=True)
pdf_sample

brand  \
item_id                             
B083VHXTQ1                     []   
B083VJ7P68                  BELEI   
B083VJ54TY                  BELEI   
B083VJ435H                  Belei   
B07SDFXTS8                  find.   
...                           ...   
B07PQQHW9V  Amazon Brand - Solimo   
B0756H8HMV           AmazonBasics   
B07PNFZ9DB  Amazon Brand - Solimo   
B07VX9RBXC           AmazonBasics   
B07D2Q4CY7           AmazonBasics   

                                                 bullet_point          color  \
item_id                                                                        
B083VHXTQ1  [With this firming eye mask, your skin will lo...            NaN   
B083VJ7P68  [Rediscover youthful looking skin with this fi...            NaN   
B083VJ54TY  [Rediscover youthful looking skin with this fi...            NaN   
B083VJ435H  [Rediscover youthful looking skin with this fi...            NaN   
B07SDFXTS8  [Pack contains: 4 bottles x 150ml, Gently and ...            NaN   
...                                                       ...            ...   
B07PQQHW9V  [A set of 20 White Board Markers in black(7), ...  multicoloured   
B0756H8HMV  [12-pack of gel ink pens pre-filled with black...          Black   
B07PNFZ9DB  [A set of 20 Permanent Markers in black(5), bl...  multicoloured   
B07VX9RBXC                                                NaN          Black   
B07D2Q4CY7  [200 sheets of loose-leaf filler paper (6-pack...            NaN   

           fabric_type                                          item_name  \
item_id                                                                     
B083VHXTQ1         NaN  Amazon Brand:Belei Moisturizing Lifting Eye Ma...   
B083VJ7P68         NaN  Amazon Brand - Belei - Lifting Sheet Mask with...   
B083VJ54TY         NaN  Amazon Brand - Belei - Full face treatment she...   
B083VJ435H         NaN  Amazon Brand - Belei - Brightening Sheet Mask ...   
B07SDFXTS8         NaN  FIND - Gentle Eye Make-Up Remover Oil Free (4 ...   
...                ...                                                ...   
B07PQQHW9V         NaN  Amazon Brand - Solimo White Board Markers Set ...   
B0756H8HMV         NaN  AmazonBasics Retractable Gel Ink Pens - Fine P...   
B07PNFZ9DB         NaN  Amazon Brand - Solimo Permanent Marker Set (20...   
B07VX9RBXC         NaN  AmazonBasics Low-Odor Dry Erase White Board Ma...   
B07D2Q4CY7         NaN  AmazonBasics College Ruled Loose Leaf Filler P...   

           item_weight model_name        product_type                   style  \
item_id                                                                         
B083VHXTQ1         NaN        NaN         ABIS_BEAUTY                      []   
B083VJ7P68         NaN        NaN         ABIS_BEAUTY                 Lifting   
B083VJ54TY         NaN        NaN         ABIS_BEAUTY     Full face treatment   
B083VJ435H         NaN        NaN         ABIS_BEAUTY             Brightening   
B07SDFXTS8        0.82        NaN         ABIS_BEAUTY  Non-Waterproof Make Up   
...                ...        ...                 ...                     ...   
B07PQQHW9V         299        NaN  WRITING_INSTRUMENT                     NaN   
B0756H8HMV    0.022443        NaN  WRITING_INSTRUMENT              Fine Point   
B07PNFZ9DB         299        NaN  WRITING_INSTRUMENT                     NaN   
B07VX9RBXC        15.4        NaN  WRITING_INSTRUMENT                     NaN   
B07D2Q4CY7    1.508692        NaN       WRITING_PAPER                     NaN   

           main_image_id                                     other_image_id  \
item_id                                                                       
B083VHXTQ1   51ZvgucquXL  [51PSBY8h8LL, 61ZcL8JqMVL, 61ZomqVpklL, 61ykRD...   
B083VJ7P68   51G7HvO94pL  [41qhq+Q55zL, 616anjMkgcL, 51cxHmZQZuL, 61NCNb...   
B083VJ54TY   51eC8v7TQVL  [51195iJBIiL, 71UsrlwqWXL, 516GHxy3KEL, 61zImT...   
B083VJ435H   51LoGumNTWL  [61V1PqjdRQL, 61AbctEZQJL, 4

In [ ]:
# pdf_sample.to_pickle("../../ABO_dataset/abo-listings-sample.pkl")
pdf_sample = pd.read_pickle("../../ABO_dataset/abo-listings-sample.pkl")
pdf_sample

brand  \
item_id                             
B083VHXTQ1                     []   
B083VJ7P68                  BELEI   
B083VJ54TY                  BELEI   
B083VJ435H                  Belei   
B07SDFXTS8                  find.   
...                           ...   
B07PQQHW9V  Amazon Brand - Solimo   
B0756H8HMV           AmazonBasics   
B07PNFZ9DB  Amazon Brand - Solimo   
B07VX9RBXC           AmazonBasics   
B07D2Q4CY7           AmazonBasics   

                                                 bullet_point          color  \
item_id                                                                        
B083VHXTQ1  [With this firming eye mask, your skin will lo...            NaN   
B083VJ7P68  [Rediscover youthful looking skin with this fi...            NaN   
B083VJ54TY  [Rediscover youthful looking skin with this fi...            NaN   
B083VJ435H  [Rediscover youthful looking skin with this fi...            NaN   
B07SDFXTS8  [Pack contains: 4 bottles x 150ml, Gently and ...            NaN   
...                                                       ...            ...   
B07PQQHW9V  [A set of 20 White Board Markers in black(7), ...  multicoloured   
B0756H8HMV  [12-pack of gel ink pens pre-filled with black...          Black   
B07PNFZ9DB  [A set of 20 Permanent Markers in black(5), bl...  multicoloured   
B07VX9RBXC                                                NaN          Black   
B07D2Q4CY7  [200 sheets of loose-leaf filler paper (6-pack...            NaN   

           fabric_type                                          item_name  \
item_id                                                                     
B083VHXTQ1         NaN  Amazon Brand:Belei Moisturizing Lifting Eye Ma...   
B083VJ7P68         NaN  Amazon Brand - Belei - Lifting Sheet Mask with...   
B083VJ54TY         NaN  Amazon Brand - Belei - Full face treatment she...   
B083VJ435H         NaN  Amazon Brand - Belei - Brightening Sheet Mask ...   
B07SDFXTS8         NaN  FIND - Gentle Eye Make-Up Remover Oil Free (4 ...   
...                ...                                                ...   
B07PQQHW9V         NaN  Amazon Brand - Solimo White Board Markers Set ...   
B0756H8HMV         NaN  AmazonBasics Retractable Gel Ink Pens - Fine P...   
B07PNFZ9DB         NaN  Amazon Brand - Solimo Permanent Marker Set (20...   
B07VX9RBXC         NaN  AmazonBasics Low-Odor Dry Erase White Board Ma...   
B07D2Q4CY7         NaN  AmazonBasics College Ruled Loose Leaf Filler P...   

           item_weight model_name        product_type                   style  \
item_id                                                                         
B083VHXTQ1         NaN        NaN         ABIS_BEAUTY                      []   
B083VJ7P68         NaN        NaN         ABIS_BEAUTY                 Lifting   
B083VJ54TY         NaN        NaN         ABIS_BEAUTY     Full face treatment   
B083VJ435H         NaN        NaN         ABIS_BEAUTY             Brightening   
B07SDFXTS8        0.82        NaN         ABIS_BEAUTY  Non-Waterproof Make Up   
...                ...        ...                 ...                     ...   
B07PQQHW9V         299        NaN  WRITING_INSTRUMENT                     NaN   
B0756H8HMV    0.022443        NaN  WRITING_INSTRUMENT              Fine Point   
B07PNFZ9DB         299        NaN  WRITING_INSTRUMENT                     NaN   
B07VX9RBXC        15.4        NaN  WRITING_INSTRUMENT                     NaN   
B07D2Q4CY7    1.508692        NaN       WRITING_PAPER                     NaN   

           main_image_id                                     other_image_id  \
item_id                                                                       
B083VHXTQ1   51ZvgucquXL  [51PSBY8h8LL, 61ZcL8JqMVL, 61ZomqVpklL, 61ykRD...   
B083VJ7P68   51G7HvO94pL  [41qhq+Q55zL, 616anjMkgcL, 51cxHmZQZuL, 61NCNb...   
B083VJ54TY   51eC8v7TQVL  [51195iJBIiL, 71UsrlwqWXL, 516GHxy3KEL, 61zImT...   
B083VJ435H   51LoGumNTWL  [61V1PqjdRQL, 61AbctEZQJL, 4

#### Language Checking

We use Google's MediaPipe language detection for Edge AI, then check non-English detections against Google Cloud language detection.

See the language detection notebook for an analysis of three different language detection models.

First, convert the relevant data from the rows to strings.

In [3]:
def row_to_text(row):
    row_filtered = row.drop(labels=['brand', 'item_weight', 'model_name', 'product_type', 'main_image_id', 'other_image_id', 'country', 'marketplace', 'domain_name', 'model_year']).dropna()
    text = []
    for item in row_filtered:
        if isinstance(item, list):
            text.extend(item)
        else:
            text.append(item)
    
    return ' '.join(text).replace('\n', ' ')

text_for_detection = [row_to_text(pdf.loc[item_id]) for item_id in tqdm(pdf.index)]

  0%|          | 0/120382 [00:00<?, ?it/s]

100%|██████████| 120382/120382 [00:30<00:00, 3976.54it/s]


Load then run the MediaPipe language detection

In [5]:
base_options = python.BaseOptions(model_asset_path="../../assets/language_detector.tflite")
options = text.LanguageDetectorOptions(base_options=base_options)
mediapipe_detector = text.LanguageDetector.create_from_options(options)

In [7]:
mediapipe_detection_results = {'languages': [], 'confidences': []}

for item in tqdm(text_for_detection):
    mediapipe_result = mediapipe_detector.detect(item).detections
    if mediapipe_result:
        mediapipe_detection_results['languages'].append(mediapipe_result[0].language_code)
        mediapipe_detection_results['confidences'].append(mediapipe_result[0].probability)
    else:
        mediapipe_detection_results['languages'].append(None)
        mediapipe_detection_results['confidences'].append(None)
        
mediapipe_detection_results_df = pd.DataFrame(mediapipe_detection_results)

100%|██████████| 120382/120382 [00:19<00:00, 6095.17it/s]


Put any non-English detections (including null results) through Google Cloud language detection.

In [8]:
indexes_for_cloud_detection = mediapipe_detection_results_df[mediapipe_detection_results_df['languages'] != 'en'].index
len(indexes_for_cloud_detection)

855

In [9]:
def detect_language_google_cloud(text: list) -> dict:
    """Detects the text's language."""
    from google.cloud import translate_v2 as translate

    translate_client = translate.Client()

    # Text can also be a sequence of strings, in which case this method
    # will return a sequence of results for each text.
    result = translate_client.detect_language(text)

    return result

In [10]:
cloud_detection_results = {'index': [], 'languages': [], 'confidences': []}

for idx in tqdm(indexes_for_cloud_detection):
    text_item = text_for_detection[idx]
    google_cloud_detection = detect_language_google_cloud(text_item)
    cloud_detection_results['index'].append(idx)
    cloud_detection_results['languages'].append(google_cloud_detection['language'])
    cloud_detection_results['confidences'].append(google_cloud_detection['confidence'])
    
cloud_detection_results_df = pd.DataFrame(cloud_detection_results).set_index('index')

100%|██████████| 855/855 [14:34<00:00,  1.02s/it]


We don't want to have to do that again, so save the cloud language detections.

In [ ]:
# cloud_detection_results_df.to_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-cloud-language-detections.pkl")
cloud_detection_results_df = pd.read_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-cloud-language-detections.pkl")
cloud_detection_results_df

languages  confidences
index                        
3             en     0.751012
172           en     0.622272
360           rw     0.397104
447      ja-Latn     0.544864
449           en     0.879083
...          ...          ...
119759        en     0.411307
119772        en     0.561561
120093        da     1.000000
120101        en     0.369890
120282        en     0.505214

[855 rows x 2 columns]

In [12]:
cloud_detection_results_df[cloud_detection_results_df['languages']!='en']

languages  confidences
index                        
360           rw     0.397104
447      ja-Latn     0.544864
646           fi     0.303385
889          war     0.414933
1050     ja-Latn     0.321378
...          ...          ...
118110        om     0.349250
118155       kha     0.344741
118822        ja     1.000000
119193        fr     0.842612
120093        da     1.000000

[310 rows x 2 columns]

All that's left is to check if the rest really can't pass as English.

In [15]:
non_eng_idxs = cloud_detection_results_df[cloud_detection_results_df['languages']!='en'].index
for idx in non_eng_idxs:
    print(f'{idx}: {text_for_detection[idx]}')

360: [Find] Amazon Collection roziumumekkisuta-ringusiruba-kuriaringu 3.5 mm kyu-bikkuzirukoniaba-pendantotye-n with Pendant jep00456 _ 110cl00jd00 
447: [Find] Amazon Collection roziumumekkisuta-ringusiruba-kuriaha-to Notebook kyu-bikkuzirukoniaharo-ringu, Size 11 
646: Cubic zirconia (CZ) [Find] Amazon Collection roziumumekkisuta-ringusiruba-kuriaasuka- & Round kattokyu-bikkuzirukoniainsa-toringusetto, Size 9 
889: Brown Napa Tan Napa Tan find. Acton Derbys, Brown Napa Tan), 6 UK Acton UK
1050: [Find] Amazon Collection 18 K Yellow Gold Silver daiyamondoraundokurasuta-buraidaruringu, Size 7 
1292: Store in a cool dry place Amazon Brand - Vedaka Popular Green Moong Split, 500g moong dal green gram lentil moong daal pulses chana dal dal green moong dal dal makhni moong dal green gram lentil moong daal pulses chana dal dal green moong dal dal makhni moong dal green gram lentil moong daal pulses chana dal dal green moong dal dal makhni moong dal green gram lentil moong daal pulses chana d

There are 310 items that not even Google Cloud language detection could detect as English. Some are obviously just using special words (e.g. Shrimp Jumbalaya). However, we must decide what to do with them. I'm going to keep some specific ones while dropping the rest.

In [13]:
idxs_to_not_drop = [5205, 7344, 9407, 16885, 18658, 22066, 25430, 26683, 28422, 35301, 43730, 46497, 54544, 64989, 67539, 74752, 86711, 87170,
                    88574, 88918, 90516, 91059, 95888, 99579, 99919, 103418, 108689, 115425, 119193, 120093]

In [25]:
idxs_to_drop = list(set(non_eng_idxs) - set(idxs_to_not_drop))
pdf = pdf.drop(pdf.index[idxs_to_drop])

In [ ]:
# pdf.to_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-verified-english.pkl")
pdf = pd.read_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-verified-english.pkl")
pdf

brand  \
item_id                             
B0857LSVB7  Amazon Brand - Solimo   
B07C5FF8QS           AmazonBasics   
B07TG425LX  Amazon Brand - Solimo   
B07LCHFZCW              Amazon Go   
B077W2YX72         Wickedly Prime   
...                           ...   
B08569DDNF  Amazon Brand - Solimo   
B0854LJBYD  Amazon Brand - Solimo   
B081HNTNK7  Amazon Brand - Solimo   
B07THSGHVP  Amazon Brand - Solimo   
B07DFWQV5V  Amazon Brand - Solimo   

                                                 bullet_point          color  \
item_id                                                                        
B0857LSVB7  [Snug fit for Samsung Galaxy J2 Ace, with perf...     Multicolor   
B07C5FF8QS  For best performance, follow the manufacturer'...            NaN   
B07TG425LX  [3D Printed Hard Back Case Mobile Cover for Hu...         Others   
B07LCHFZCW                                                NaN            NaN   
B077W2YX72  [One 44-ounce plastic tub, Contains wheat, pea...            NaN   
...                                                       ...            ...   
B08569DDNF  [Snug fit for Samsung Galaxy A2 Core, with per...     Multicolor   
B0854LJBYD  [Snug fit for iKall K200, with perfect cut-out...     Multicolor   
B081HNTNK7  [Snug fit for Mobile, with perfect cut-outs fo...  multi-colored   
B07THSGHVP  [3D Printed Hard Back Case Mobile Cover for Xi...         Others   
B07DFWQV5V  [Non-stick heating plates are double Teflon co...          Black   

           fabric_type                                          item_name  \
item_id                                                                     
B0857LSVB7         NaN  Amazon Brand - Solimo Designer Lion UV Printed...   
B07C5FF8QS         NaN  AmazonBasics High Mileage Motor Oil, Synthetic...   
B07TG425LX         NaN  Amazon Brand - Solimo Designer Semi Circle Tex...   
B07LCHFZCW         NaN   Amazon Kitchen, Cajun Style Potato Salad, 7.2 oz   
B077W2YX72         NaN  Wickedly Prime Peanut Butter-Filled Pretzels, ...   
...                ...                                                ...   
B08569DDNF         NaN  Amazon Brand - Solimo Designer Chinnese Yin an...   
B0854LJBYD         NaN  Amazon Brand - Solimo Designer Teddy Bear UV P...   
B081HNTNK7         NaN  Amazon Brand - Solimo Designer Alphabet Patter...   
B07THSGHVP         NaN  Amazon Brand - Solimo Designer Simple White Ra...   
B07DFWQV5V         NaN  Amazon Brand - Solimo Tandoor Grill (1500W, Bl...   

           item_weight              model_name                  product_type  \
item_id                                                                        
B0857LSVB7          50   Samsung Galaxy J2 Ace           CELLULAR_PHONE_CASE   
B07C5FF8QS         NaN                     NaN                      AUTO_OIL   
B07TG425LX          50          Huawei P9 lite           CELLULAR_PHONE_CASE   
B07LCHFZCW         NaN                     NaN                       GROCERY   
B077W2YX72         2.8                     NaN                       GROCERY   
...                ...                     ...                           ...   
B08569DDNF          50  Samsung Galaxy A2 Core           CELLULAR_PHONE_CASE   
B0854LJBYD          50              iKall K200           CELLULAR_PHONE_CASE   
B081HNTNK7          50            Vivo V17 Pro           CELLULAR_PHONE_CASE   
B07THSGHVP          50    Xiaomi Redmi Y1 Lite           CELLULAR_PHONE_CASE   
B07DFWQV5V         NaN              SOL_KA_103  COUNTERTOP_GRIDDLE_APPLIANCE   

                                     style main_image_id  \
item_id                                                    
B0857LSVB7                             NaN   81-DuD5XzmL   
B07C5FF8QS  High Mileage - Synthetic Blend   81YCp3dcurL   
B07TG425LX                             NaN   71ZoXfUr-sL   
B07LCHFZCW                             NaN   61bCuBtuZ4L   
B077W2YX72                             NaN   71QbaO8qZIL   
...                                    ...

#### Fix Product Type Naming

The product types tend to have underscores or nothing separating the words at all. For them to be useful, that must be fixed.

In [28]:
pdf['product_type'].nunique()

527

In [29]:
pdf['product_type'].unique()

array(['CELLULAR_PHONE_CASE', 'AUTO_OIL', 'GROCERY', 'SALWAR_SUIT_SET',
       'SHOES', 'RECREATION_BALL', 'CHAIR', 'ACCESSORY',
       'HEALTH_PERSONAL_CARE', 'BATTERY', 'NUTS',
       'HOME_LIGHTING_AND_LAMPS', 'HOME', 'CHARGING_ADAPTER',
       'HOME_FURNITURE_AND_DECOR', 'VIDEO_GAME_ACCESSORIES',
       'FURNITURE_COVER', 'WALL_ART', 'BED', 'SOFA', 'BRACELET', 'HERB',
       'HOME_BED_AND_BATH', 'RUG', 'BABY_PRODUCT', 'PLANTER',
       'STOOL_SEATING', 'TABLE', 'BACKPACK', 'SAUCE',
       'COMPUTER_INPUT_DEVICE', 'DAIRY_BASED_DRINK', 'SNACK_MIX',
       'VITAMIN', 'SHELF', 'PET_SUPPLIES', 'SANDAL', 'TOOLS',
       'LIGHT_BULB', 'FINERING', 'CAKE', 'FINENECKLACEBRACELETANKLET',
       'JANITORIAL_SUPPLY', 'COFFEE', 'SUITCASE', 'FINEEARRING',
       'EYEWEAR', 'INSTRUMENT_PARTS_AND_ACCESSORIES', 'LABEL', 'BREAD',
       'NOODLE', 'BEAUTY', 'SAFETY_SUPPLY', 'NUTRITIONAL_SUPPLEMENT',
       'LAMP', 'HARDWARE', 'HERBAL_SUPPLEMENT', 'LIGHT_FIXTURE',
       'STORAGE_BINDER', 'FLASH_MEMORY

In [30]:
pdf['product_type'] = pdf['product_type'].str.replace('_', ' ')
pdf['product_type'].unique()

array(['CELLULAR PHONE CASE', 'AUTO OIL', 'GROCERY', 'SALWAR SUIT SET',
       'SHOES', 'RECREATION BALL', 'CHAIR', 'ACCESSORY',
       'HEALTH PERSONAL CARE', 'BATTERY', 'NUTS',
       'HOME LIGHTING AND LAMPS', 'HOME', 'CHARGING ADAPTER',
       'HOME FURNITURE AND DECOR', 'VIDEO GAME ACCESSORIES',
       'FURNITURE COVER', 'WALL ART', 'BED', 'SOFA', 'BRACELET', 'HERB',
       'HOME BED AND BATH', 'RUG', 'BABY PRODUCT', 'PLANTER',
       'STOOL SEATING', 'TABLE', 'BACKPACK', 'SAUCE',
       'COMPUTER INPUT DEVICE', 'DAIRY BASED DRINK', 'SNACK MIX',
       'VITAMIN', 'SHELF', 'PET SUPPLIES', 'SANDAL', 'TOOLS',
       'LIGHT BULB', 'FINERING', 'CAKE', 'FINENECKLACEBRACELETANKLET',
       'JANITORIAL SUPPLY', 'COFFEE', 'SUITCASE', 'FINEEARRING',
       'EYEWEAR', 'INSTRUMENT PARTS AND ACCESSORIES', 'LABEL', 'BREAD',
       'NOODLE', 'BEAUTY', 'SAFETY SUPPLY', 'NUTRITIONAL SUPPLEMENT',
       'LAMP', 'HARDWARE', 'HERBAL SUPPLEMENT', 'LIGHT FIXTURE',
       'STORAGE BINDER', 'FLASH MEMORY

Items to fix manually:
* FINERING
* FINENECKLACEBRACELETANKLET
* FINEEARRING
* ABIS BEAUTY
* ABIS LAWN AND GARDEN
* ABIS DRUGSTORE
* ABIS BOOK
* FASHIONNECKLACEBRACELETANKLET
* ABIS KITCHEN
* ABIS HOME IMPROVEMENT
* FINEOTHER
* FASHIONEARRING
* SHOWERHEAD
* ABIS ELECTRONICS
* ABIS VIDEO GAMES
* FASHIONOTHER

In [31]:
pdf.loc[pdf['product_type'] == 'FINERING', 'product_type'] = 'FINE RING'
pdf.loc[pdf['product_type'] == 'FINENECKLACEBRACELETANKLET', 'product_type'] = 'FINE NECKLACE BRACELET ANKLET'
pdf.loc[pdf['product_type'] == 'FINEEARRING', 'product_type'] = 'FINE EARRING'
pdf.loc[pdf['product_type'] == 'FASHIONNECKLACEBRACELETANKLET', 'product_type'] = 'FASHION NECKLACE BRACELET ANKLET'
pdf.loc[pdf['product_type'] == 'FINEOTHER', 'product_type'] = 'FINE OTHER'
pdf.loc[pdf['product_type'] == 'FASHIONEARRING', 'product_type'] = 'FASHION EARRING'
pdf.loc[pdf['product_type'] == 'SHOWERHEAD', 'product_type'] = 'SHOWER HEAD'
pdf.loc[pdf['product_type'] == 'FASHIONOTHER', 'product_type'] = 'FASHION OTHER'
pdf['product_type'] = pdf['product_type'].str.replace('ABIS ', '')

In [32]:
pdf['product_type'].unique()

array(['CELLULAR PHONE CASE', 'AUTO OIL', 'GROCERY', 'SALWAR SUIT SET',
       'SHOES', 'RECREATION BALL', 'CHAIR', 'ACCESSORY',
       'HEALTH PERSONAL CARE', 'BATTERY', 'NUTS',
       'HOME LIGHTING AND LAMPS', 'HOME', 'CHARGING ADAPTER',
       'HOME FURNITURE AND DECOR', 'VIDEO GAME ACCESSORIES',
       'FURNITURE COVER', 'WALL ART', 'BED', 'SOFA', 'BRACELET', 'HERB',
       'HOME BED AND BATH', 'RUG', 'BABY PRODUCT', 'PLANTER',
       'STOOL SEATING', 'TABLE', 'BACKPACK', 'SAUCE',
       'COMPUTER INPUT DEVICE', 'DAIRY BASED DRINK', 'SNACK MIX',
       'VITAMIN', 'SHELF', 'PET SUPPLIES', 'SANDAL', 'TOOLS',
       'LIGHT BULB', 'FINE RING', 'CAKE', 'FINE NECKLACE BRACELET ANKLET',
       'JANITORIAL SUPPLY', 'COFFEE', 'SUITCASE', 'FINE EARRING',
       'EYEWEAR', 'INSTRUMENT PARTS AND ACCESSORIES', 'LABEL', 'BREAD',
       'NOODLE', 'BEAUTY', 'SAFETY SUPPLY', 'NUTRITIONAL SUPPLEMENT',
       'LAMP', 'HARDWARE', 'HERBAL SUPPLEMENT', 'LIGHT FIXTURE',
       'STORAGE BINDER', 'FLASH M

In [33]:
pdf.to_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-english-fixed-product-type.pkl")
pdf = pd.read_pickle("../../../Capstone Project - ShopTalk/ABO_dataset/abo-listings-english-fixed-product-type.pkl")
pdf

brand  \
item_id                             
B0857LSVB7  Amazon Brand - Solimo   
B07C5FF8QS           AmazonBasics   
B07TG425LX  Amazon Brand - Solimo   
B07LCHFZCW              Amazon Go   
B077W2YX72         Wickedly Prime   
...                           ...   
B08569DDNF  Amazon Brand - Solimo   
B0854LJBYD  Amazon Brand - Solimo   
B081HNTNK7  Amazon Brand - Solimo   
B07THSGHVP  Amazon Brand - Solimo   
B07DFWQV5V  Amazon Brand - Solimo   

                                                 bullet_point          color  \
item_id                                                                        
B0857LSVB7  [Snug fit for Samsung Galaxy J2 Ace, with perf...     Multicolor   
B07C5FF8QS  For best performance, follow the manufacturer'...            NaN   
B07TG425LX  [3D Printed Hard Back Case Mobile Cover for Hu...         Others   
B07LCHFZCW                                                NaN            NaN   
B077W2YX72  [One 44-ounce plastic tub, Contains wheat, pea...            NaN   
...                                                       ...            ...   
B08569DDNF  [Snug fit for Samsung Galaxy A2 Core, with per...     Multicolor   
B0854LJBYD  [Snug fit for iKall K200, with perfect cut-out...     Multicolor   
B081HNTNK7  [Snug fit for Mobile, with perfect cut-outs fo...  multi-colored   
B07THSGHVP  [3D Printed Hard Back Case Mobile Cover for Xi...         Others   
B07DFWQV5V  [Non-stick heating plates are double Teflon co...          Black   

           fabric_type                                          item_name  \
item_id                                                                     
B0857LSVB7         NaN  Amazon Brand - Solimo Designer Lion UV Printed...   
B07C5FF8QS         NaN  AmazonBasics High Mileage Motor Oil, Synthetic...   
B07TG425LX         NaN  Amazon Brand - Solimo Designer Semi Circle Tex...   
B07LCHFZCW         NaN   Amazon Kitchen, Cajun Style Potato Salad, 7.2 oz   
B077W2YX72         NaN  Wickedly Prime Peanut Butter-Filled Pretzels, ...   
...                ...                                                ...   
B08569DDNF         NaN  Amazon Brand - Solimo Designer Chinnese Yin an...   
B0854LJBYD         NaN  Amazon Brand - Solimo Designer Teddy Bear UV P...   
B081HNTNK7         NaN  Amazon Brand - Solimo Designer Alphabet Patter...   
B07THSGHVP         NaN  Amazon Brand - Solimo Designer Simple White Ra...   
B07DFWQV5V         NaN  Amazon Brand - Solimo Tandoor Grill (1500W, Bl...   

           item_weight              model_name                  product_type  \
item_id                                                                        
B0857LSVB7          50   Samsung Galaxy J2 Ace           CELLULAR PHONE CASE   
B07C5FF8QS         NaN                     NaN                      AUTO OIL   
B07TG425LX          50          Huawei P9 lite           CELLULAR PHONE CASE   
B07LCHFZCW         NaN                     NaN                       GROCERY   
B077W2YX72         2.8                     NaN                       GROCERY   
...                ...                     ...                           ...   
B08569DDNF          50  Samsung Galaxy A2 Core           CELLULAR PHONE CASE   
B0854LJBYD          50              iKall K200           CELLULAR PHONE CASE   
B081HNTNK7          50            Vivo V17 Pro           CELLULAR PHONE CASE   
B07THSGHVP          50    Xiaomi Redmi Y1 Lite           CELLULAR PHONE CASE   
B07DFWQV5V         NaN              SOL_KA_103  COUNTERTOP GRIDDLE APPLIANCE   

                                     style main_image_id  \
item_id                                                    
B0857LSVB7                             NaN   81-DuD5XzmL   
B07C5FF8QS  High Mileage - Synthetic Blend   81YCp3dcurL   
B07TG425LX                             NaN   71ZoXfUr-sL   
B07LCHFZCW                             NaN   61bCuBtuZ4L   
B077W2YX72                             NaN   71QbaO8qZIL   
...                                    ...

#### Map image IDs to image files

In [ ]:
image_meta_path = '../../../Capstone Project - ShopTalk/ABO_dataset/images/metadata/images.csv'
image_meta_df = pd.read_csv(image_meta_path).set_index('image_id')
image_meta_df

height  width             path
image_id                                   
010-mllS7JL     106    106  14/14fe8812.jpg
01dkn0Gyx0L     122    122  da/daab0cad.jpg
01sUPg0387L     111    111  d2/d2daaae9.jpg
1168jc-5r1L     186    186  3a/3a4e88e6.jpg
11RUV5Fs65L      30    500  d9/d91ab9cf.jpg
...             ...    ...              ...
B1zv8OpTkBS    2560   2560  6d/6d49d130.jpg
B1zwflWhPIS    2560   2560  b1/b163e0ea.jpg
C1lf45DhhRS    2560   2560  a1/a116d9d1.jpg
C1pEt6jBLiS    2560   2560  9c/9c3e1158.jpg
C1sCYAz8NAS    2560   2560  cf/cf112e38.jpg

[398212 rows x 3 columns]

Check that all the image IDs map to an image

In [ ]:
# Previously did not check for nulls. Fixed.
missing_images = 0
all_image_paths = set()
for main_image_id in pdf['main_image_id']:
    if not pd.isna(main_image_id):
        if main_image_id in image_meta_df.index:
            all_image_paths.add(image_meta_df.loc[main_image_id]['path'])
        else:
            missing_images += 1
for other_image_id_list in pdf['other_image_id']:
    if isinstance(other_image_id_list, list):
        for other_image_id in other_image_id_list:
            if other_image_id in image_meta_df.index:
                all_image_paths.add(image_meta_df.loc[other_image_id]['path'])
            else:
                missing_images += 1
print(f'missing images = {missing_images}')
print(f'matched images = {len(all_image_paths)}')

missing images = 0
matched images = 326890
